File: CFD_Principles_and_Applications_Blazek/BookCode/C++/unstructured_2d/src/fluxRoe1.cpp
/// @file fluxRoe1.cpp
///
/// Computation of the convective fluxes based on flux averages.
///
//*****************************************************************************
//
//  (c) J. Blazek, CFD Consulting & Analysis, www.cfd-ca.de
//  Created February 15, 2014
//  Last modification: July 7, 2014
// 
//=============================================================================
//
//  This program is free software; you can redistribute it and/or
//  modify it under the terms of the GNU General Public License
//  as published by the Free Software Foundation; either version 2
//  of the License, or (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//
//*****************************************************************************

#include "spaceDiscr.h"

/// Computes the convective fluxes using average of fluxes at control volume's
/// faces. The left and right fluxes are computed directly from values at nodes
/// "i" and "j". This is sufficient for a 1st-order Roe scheme.
///
/// @param geometry    geometrical data
/// @param fluidProps  fluid properties
///
void SpaceDiscr::FluxRoe1( const Geometry &geometry, const FluidProps &fluidProps )
{
  int  i, j, ie;
  REAL rhl, rhr, qsl, qsr, pav;
  REAL fc[4];

  CONSVARS       *cv   = fluidProps.cv;
  DEPVARS        *dv   = fluidProps.dv;
  NODE           *sij  = geometry.sij;
  Geometry::EDGE *edge = geometry.edge;

  // initialize residual by adding artificial dissipation

  for (i=0; i<geometry.nNodes; i++)
  {
    rhs[i].dens = -diss[i].dens;
    rhs[i].xmom = -diss[i].xmom;
    rhs[i].ymom = -diss[i].ymom;
    rhs[i].ener = -diss[i].ener;
  }

  // average of fluxes

  for (ie=0; ie<geometry.nEdges; ie++)
  {
    i = edge[ie].i;
    j = edge[ie].j;

    // left and right rho*H, V*n

    rhl = dv[i].press + cv[i].ener;
    qsl = (cv[i].xmom*sij[ie].x+cv[i].ymom*sij[ie].y)/cv[i].dens;

    rhr = dv[j].press + cv[j].ener;
    qsr = (cv[j].xmom*sij[ie].x+cv[j].ymom*sij[ie].y)/cv[j].dens;

    // fluxes

    pav   = 0.5*(dv[i].press+dv[j].press);
    fc[0] = 0.5*(qsl*cv[i].dens+qsr*cv[j].dens);
    fc[1] = 0.5*(qsl*cv[i].xmom+qsr*cv[j].xmom) + sij[ie].x*pav;
    fc[2] = 0.5*(qsl*cv[i].ymom+qsr*cv[j].ymom) + sij[ie].y*pav;
    fc[3] = 0.5*(qsl*rhl       +qsr*rhr);

    rhs[i].dens += fc[0];
    rhs[i].xmom += fc[1];
    rhs[i].ymom += fc[2];
    rhs[i].ener += fc[3];

    rhs[j].dens -= fc[0];
    rhs[j].xmom -= fc[1];
    rhs[j].ymom -= fc[2];
    rhs[j].ener -= fc[3];
  }

  // treatment of solid walls

  FluxWalls( geometry,fluidProps );
}

File: CFD_Principles_and_Applications_Blazek/BookCode/Fortran/unstructured_2d/src/fluxRoe1.f90
!> @file fluxRoe1.f90
!!
!! Computation of the convective fluxes based on flux averages.
!
! *****************************************************************************
!
!  (c) J. Blazek, CFD Consulting & Analysis, www.cfd-ca.de
!  Created February 25, 2014
!  Last modification: June 3, 2014
!
! *****************************************************************************
!
!  This program is free software; you can redistribute it and/or
!  modify it under the terms of the GNU General Public License
!  as published by the Free Software Foundation; either version 2
!  of the License, or (at your option) any later version.
!
!  This program is distributed in the hope that it will be useful,
!  but WITHOUT ANY WARRANTY; without even the implied warranty of
!  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
!  GNU General Public License for more details.
!
!  You should have received a copy of the GNU General Public License
!  along with this program; if not, write to the Free Software
!  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
!
! *****************************************************************************

!> Computes the convective fluxes using average of fluxes at control volume's
!! faces. The left and right fluxes are computed directly from values at nodes
!! "i" and "j". This is sufficient for a 1st-order Roe scheme.
!!
subroutine FluxRoe1

  use ModDataTypes
  use ModGeometry
  use ModNumerics
  use ModPhysics
  use ModInterfaces, only : FluxWalls
  implicit none

! local variables
  integer     :: i, j, ie
  real(rtype) :: rhl, rhr, qsl, qsr, pav
  real(rtype) :: fc(4)

! *****************************************************************************
! initialize residual by adding artificial dissipation

  do i=1,nnodes
    rhs(1,i) = -diss(1,i)
    rhs(2,i) = -diss(2,i)
    rhs(3,i) = -diss(3,i)
    rhs(4,i) = -diss(4,i)
  enddo

! average of fluxes

  do ie=1,nedges
    i = edge(1,ie)
    j = edge(2,ie)

! - left and right rho*H, V*n

    rhl = dv(1,i) + cv(4,i)
    qsl = (cv(2,i)*sij(1,ie)+cv(3,i)*sij(2,ie))/cv(1,i)

    rhr = dv(1,j) + cv(4,j)
    qsr = (cv(2,j)*sij(1,ie)+cv(3,j)*sij(2,ie))/cv(1,j)

! - fluxes

    pav   = 0.5D0*(dv(1,i)+dv(1,j))
    fc(1) = 0.5D0*(qsl*cv(1,i)+qsr*cv(1,j))
    fc(2) = 0.5D0*(qsl*cv(2,i)+qsr*cv(2,j)) + sij(1,ie)*pav
    fc(3) = 0.5D0*(qsl*cv(3,i)+qsr*cv(3,j)) + sij(2,ie)*pav
    fc(4) = 0.5D0*(qsl*rhl    +qsr*rhr    )

    rhs(1,i) = rhs(1,i) + fc(1)
    rhs(2,i) = rhs(2,i) + fc(2)
    rhs(3,i) = rhs(3,i) + fc(3)
    rhs(4,i) = rhs(4,i) + fc(4)

    rhs(1,j) = rhs(1,j) - fc(1)
    rhs(2,j) = rhs(2,j) - fc(2)
    rhs(3,j) = rhs(3,j) - fc(3)
    rhs(4,j) = rhs(4,j) - fc(4)
  enddo

! treatment of solid walls

  call FluxWalls

end subroutine FluxRoe1
